tree.go

```diff
@@ -683,62 +685,86 @@ func (n *InternalNode) Commit() *Point {
 	internalNodeLevels := make([][]*InternalNode, StemSize)
 	n.fillLevels(internalNodeLevels)
 
-	points := make([]*Point, 0, 1024)
-	cowIndexes := make([]int, 0, 1024)
-	poly := make([]Fr, NodeWidth)
 	for level := len(internalNodeLevels) - 1; level >= 0; level-- {
 		nodes := internalNodeLevels[level]
 		if len(nodes) == 0 {
 			continue
 		}
-		points = points[:0]
-		cowIndexes = cowIndexes[:0]
-
-		// For each internal node, we collect in `points` all the ones we need to map to a field element.
-		// That is, for each touched children in a node, we collect the old and new commitment to do the diff updating
-		// later.
-		for _, node := range nodes {
-			for idx, nodeChildComm := range node.cow {
-				points = append(points, nodeChildComm)
-				points = append(points, node.children[idx].Commitment())
-				cowIndexes = append(cowIndexes, int(idx))
+
+		minBatchSize := 4
+		if len(nodes) <= minBatchSize {
+			commitNodesAtLevel(nodes)
+		} else {
+			var wg sync.WaitGroup
+			numBatches := runtime.NumCPU()
+			batchSize := (len(nodes) + numBatches - 1) / numBatches
+			if batchSize < minBatchSize {
+				batchSize = minBatchSize
+			}
+			for i := 0; i < len(nodes); i += batchSize {
+				start := i
+				end := i + batchSize
+				if end > len(nodes) {
+					end = len(nodes)
+				}
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					commitNodesAtLevel(nodes[start:end])
+				}()
 			}
+			wg.Wait()
 		}
+	}
+	return n.commitment
+}
 
-		// We generate `frs` which will contain the result for each element in `points`.
-		frs := make([]*Fr, len(points))
-		for i := range frs {
-			frs[i] = &Fr{}
+func commitNodesAtLevel(nodes []*InternalNode) {
+	points := make([]*Point, 0, 1024)
+	cowIndexes := make([]int, 0, 1024)
+
+	// For each internal node, we collect in `points` all the ones we need to map to a field element.
+	// That is, for each touched children in a node, we collect the old and new commitment to do the diff updating
+	// later.
+	for _, node := range nodes {
+		for idx, nodeChildComm := range node.cow {
+			points = append(points, nodeChildComm)
+			points = append(points, node.children[idx].Commitment())
+			cowIndexes = append(cowIndexes, int(idx))
 		}
+	}
 
-		// Do a single batch calculation for all the points in this level.
-		toFrMultiple(frs, points)
+	// We generate `frs` which will contain the result for each element in `points`.
+	frs := make([]*Fr, len(points))
+	for i := range frs {
+		frs[i] = &Fr{}
+	}
 
-		// We calculate the difference between each (new commitment - old commitment) pair, and store it
-		// in the same slice to avoid allocations.
-		for i := 0; i < len(frs); i += 2 {
-			frs[i/2].Sub(frs[i+1], frs[i])
-		}
-		// Now `frs` have half of the elements, and these are the Frs differences to update commitments.
-		frs = frs[:len(frs)/2]
+	// Do a single batch calculation for all the points in this level.
+	toFrMultiple(frs, points)
```

Comment:
So this `toFrMultiple` before was a single call per level; now by the nature of the work-splitting we're doing N `toFrMultiple` calls. So that means some extra inversions in total, but using all cores which is much faster.

Reply:
is there a known cut-off point after which it makes sense not to do too many `toFrMultiple` or is it already faster with 2 cores?